boa/src/builtins/function/mod.rs

```diff
@@ -1,134 +1,318 @@
-//! This module implements the global `Function` object.
+//! This module implements the global `Function` object as well as creates Native Functions.
 //!
-//! `Every JavaScript `function` is actually a `Function` object.
+//! Objects wrap `Function`s and expose them via call/construct slots.
+//!
+//! `The `Function` object is used for matching text with a pattern.
 //!
 //! More information:
 //!  - [ECMAScript reference][spec]
 //!  - [MDN documentation][mdn]
 //!
-//! [spec]: https://tc39.es/ecma262/#sec-function-object
+//! [spec]: https://tc39.es/ecma262/#sec-function-objects
 //! [mdn]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function
 
-#[cfg(test)]
-mod tests;
-
 use crate::{
     builtins::{
-        object::{internal_methods_trait::ObjectInternalMethods, Object},
+        array,
+        object::{Object, ObjectInternalMethods, ObjectKind, PROTOTYPE},
         property::Property,
         value::{to_value, ResultValue, Value, ValueData},
     },
-    exec::Interpreter,
+    environment::lexical_environment::{new_function_environment, Environment},
+    exec::Executor,
     syntax::ast::node::{FormalParameter, Node},
+    Interpreter,
 };
-use gc::{custom_trace, Gc};
+use gc::{unsafe_empty_trace, Gc, Trace as TraceTrait};
 use gc_derive::{Finalize, Trace};
 use std::fmt::{self, Debug};
-use std::ops::Deref;
 
-/// fn(this, arguments, ctx)
-pub type NativeFunctionData = fn(&Value, &[Value], &mut Interpreter) -> ResultValue;
+/// _fn(this, arguments, ctx) -> ResultValue_ - The signature of a built-in function
+pub type NativeFunctionData = fn(&mut Value, &[Value], &mut Interpreter) -> ResultValue;
 
-/// A Javascript `Function` object instance.
-///
-/// A member of the Object type that may be invoked as a subroutine
+/// Sets the ConstructorKind
+#[derive(Debug, Copy, Clone)]
+pub enum ConstructorKind {
+    Base,
+    Derived,
+}
+/// Defines how this references are interpreted within the formal parameters and code body of the function.
 ///
-/// In our implementation, Function is extending Object by holding an object field which some extra data
+/// Arrow functions don't define a `this` and thus are lexical, `function`s do define a this and thus are NonLexical
 #[derive(Trace, Finalize, Debug, Clone)]
-pub enum Function {
-    /// A native javascript function
-    NativeFunc(NativeFunction),
-    /// A regular javascript function
-    RegularFunc(RegularFunction),
+pub enum ThisMode {
+    Lexical,
+    NonLexical,
 }
 
-/// Represents a regular javascript function in memory.
-#[derive(Trace, Finalize, Debug, Clone)]
-pub struct RegularFunction {
-    /// The fields associated with the function
-    pub object: Object,
-    /// This function's expression
-    pub node: Node,
-    /// The argument declarations of the function
-    pub args: Vec<Node>,
+/// FunctionBody is specific to this interpreter, it will either be Rust code or JavaScript code (AST Node)
+#[derive(Clone, Finalize)]
```

Comment:
It's defined on line 76.
We can't derive it because `Gc` doesn't natively support fn(_, _, _).
Seeing that FunctionBody doesn't hold other GC values its safe to mark it as an unsafe_empty_trace

Reply:
I see, it was implemented in a different place and I missed it! maybe it would make sense to move it closer to the definition of the `struct`

Reply:
Its just underneath now, after  Debug for FunctionBody